packages/orm/src/FlushProxy.ts

```diff
@@ -0,0 +1,120 @@
+import { OneToManyCollection } from "./collections/OneToManyCollection";
+import { ManyToManyCollection } from "./collections/ManyToManyCollection";
+import { ManyToOneReference } from "./collections/ManyToOneReference";
+import { Entity, EntityManager } from "./EntityManager";
+import * as util from "util";
+
+// These proxies are passed into hooks during a flush to enable changes mid flush
+
+type FlushProxy<T> = T & { __flushSecret: number; __proxyTarget: T };
+type Loader = OneToManyCollection<any, any> | ManyToManyCollection<any, any> | ManyToOneReference<any, any, any>;
+
+export function isFlushProxy<T>(objectOrProxy: T): objectOrProxy is FlushProxy<T> {
+  return util.types.isProxy(objectOrProxy) && (objectOrProxy as any).__flushSecret !== undefined;
+}
+
+export function getTargetFromMaybeFlushProxy<T>(objectOrProxy: T | FlushProxy<T>): T {
+  return isFlushProxy(objectOrProxy) ? objectOrProxy.__proxyTarget : objectOrProxy;
+}
+
+function proxyLoaderResult(result: any, flushSecret: number) {
+  if (Array.isArray(result)) {
+    return result.map((entity) => proxyEntity(entity, flushSecret));
+  } else if (typeof result === "object") {
+    return proxyEntity(result, flushSecret);
+  }
+
+  return result;
+}
+
+function createFlushProxy<T extends object>(
+  target: T,
+  flushSecret: number,
+  block: (target: T, property: PropertyKey, receiver: any) => any,
+): FlushProxy<T> {
+  if (isFlushProxy(target)) {
+    return target;
+  }
+
+  return new Proxy(target, {
+    get(target, property, receiver) {
+      if (property === "__flushSecret") {
+        return flushSecret;
+      }
+
+      if (property === "__proxyTarget") {
+        return target;
+      }
+
+      return block(target, property, receiver);
+    },
+  }) as FlushProxy<T>;
+}
+
+function proxyLoader(loader: Loader, entity: Entity, flushSecret: number): FlushProxy<Loader> {
+  return createFlushProxy(loader, flushSecret, (loader, property, receiver) => {
+    if (property === "load") {
+      return function () {
+        return loader.load().then((result) => proxyLoaderResult(result, flushSecret));
+      };
+    }
+
+    if (property === "get") {
+      return proxyLoaderResult(loader.get, flushSecret);
+    }
+
+    if (property === "entity") {
+      return proxyEntity(entity, flushSecret);
+    }
+
+    return Reflect.get(loader, property, receiver);
+  });
+}
+
+export function proxyEntityManager(em: EntityManager, flushSecret: number): FlushProxy<EntityManager> {
```

Comment:
Hrm. Could we remove all of the `proxyLoaderResult` / `proxyEntity` / `proxyEntityManager` specific logic? I'm thinking ...probably? Given that demo of "if you keep returning a proxy with your same handler, and you can see anything they access through `this` N-levels deep"...

I could see needing to inspect the `Reflect.get` return value to handle a few cases like arrays and promises...but other than that, do we care about not proxying a few extra methods? (I.e. AFAICT the main goal of having these per-type proxies is to a) know about promises (but could check that generically via `then` detection and b) not proxying things that we don't think we need to).

I suppose if you wanted to early return from a proxy, what about proxying every method, but then stopping if the return value is not an array/promise/Entity/EntityManager/Reference?

Reply:
(re-articulating for my own benefit... ideally the `block` param of `createFlushProxy` would go away, and it could generically handle any result that it traverses across, by defaulting to always re-proxying it's own handler again/again at each step, and ideally special casing as little as possible, i.e. maybe only promises? and that is it?)

Reply:
The idea being both in terms of simplification/less LOC but also robustness against future changes to the entity/EM/loader APIs.